### ISSUE TITLE
Question about Kiam and hostNetwork

### ISSUE THREAD

**Issue by gillg** [aws/aws-node-termination-handler#247]
Hi,

I don't understand why we should add kiam whitelist provided in #187 
In fact, in DaemonSet charts pod is configured with `hostNetwork: true`, so kiam should not intercept metadata call on host network interface.
But I also ask me why this pod needs "hostNetwork" permissions ? 
It only needs metadata api access and kubernetes API... no ?

Regards

**Comment by bwagner5** [comment]
@gillg  The NTH pods need hostNetwork to work w/ IMDSv2 which by default only allows 1 network hop. 

You may be correct about kiam not intercepting calls to when the pod is using HostNetworking.  

Looks like the default setup is to put the iptables intercept rule on the Docker0 bridge interface:  
https://github.com/uswitch/kiam/blob/0fe1a077e3a408cf47dea25ac83e14e367fb0bb9/cmd/kiam/agent.go#L64

@leosunmo do you have any insight on this? Is it common for users to override the interface to `eth0` or the like that would intercept requests to metadata from hostNetworking pods?

**Comment by bwagner5** [comment]
I'm pretty sure we can remove the kiam steps from the readme, thanks for bringing this up @gillg

**Comment by gillg** [comment]
To be honest I think it's a good basic whitelist for kiam, but not mandatory for this project if kiam is correctly configured.

Last point, I'm not familiar with imdsv2 but what is changing ? If you use kiam agent proxy you not really use this feature ?

**Comment by bwagner5** [comment]
NTH defaults to use IMDSv2 and will fallback to IMDSv1 if it can't reach v2. The metadata service is the same in both versions (same paths and responses), but IMDSv2 requires a token and has a limit of 1 hop at the IP level.  I'm not sure how the kiam agent proxy works with IMDSv2, it may work, but haven't tried it. 

"By default, the response to PUT requests has a response hop limit (time to live) of 1 at the IP protocol level. You can adjust the hop limit using the modify-instance-metadata-options command if you need to make it larger. For example, you might need a larger hop limit for backward compatibility with container services running on the instance. For more information, see modify-instance-metadata-options in the AWS CLI Command Reference."
- https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html

Here's a good blog post on the reasoning behind IMDSv2: https://aws.amazon.com/blogs/security/defense-in-depth-open-firewalls-reverse-proxies-ssrf-vulnerabilities-ec2-instance-metadata-service/

**Comment by gillg** [comment]
Thanks for you blog article it was very interesting !
So, in my opinion (need some tests), if you not use kiam, you should use hostNetwork, but with kiam probably not.
It work as a kind of mock of IMDS. You install one agent (Daemonset) which capture all traffic to 169.254.169.254 with a dnat firewalling rule, to forward request to itself. It checks if requester pod is permit to assume any role (or if route is whitelisted), then it delegates IMDS call to a server pod (on the same or another node). This server uses hostNetwork and sould be authorized to assume role asked by pod. It makes cache of credentials for assumed roles if possible, to avoid too much IMDS calls.
So, the only real credentials fetch is done by the server pod. 
Except if I misunderstood, this scenario should work nevertheless the IMDS version...

**Comment by leosunmo** [comment]
Hey, I just got to this now.
I had a colleague run this through and they found no issues with KIAM + NTH, so I must have not used `hostNetwork: true` or something when I first ran it, or something else went wrong.
So to me it would make sense to remove the docs since most people would use the Helm chart, which hardcodes `hostNetwork: true`, so the docs are of no use!

**Comment by gillg** [comment]
For your information I opened an issue about hostNetwork and more generaly about security of helm chart. https://github.com/aws/eks-charts/issues/238

**Comment by leosunmo** [comment]
I'll link this here so we have it in this discussion as well.
https://github.com/aws/eks-charts/issues/238#issuecomment-687523076

I tested without `hostNetwork: true` and with the Kiam allowlist described in the docs, and it works fine. I verified that it is normally blocked by Kiam, and immediately got it working when the allowlist was added. I think we can probably drop the `hostNetwork: true` from the Helm chart, or make it an opt-in in case there's some special cases where your overlay network prevents you from reaching the metadata endpoint.
